client/allocrunner/networking_cni.go

```diff
@@ -171,6 +172,16 @@ func (c *cniNetworkConfigurator) Setup(ctx context.Context, alloc *structs.Alloc
 		cniArgs[ConsulIPTablesConfigEnvVar] = string(iptablesCfg)
 	}
 
+	if !created {
+		err := c.cni.Check(ctx, alloc.ID, spec.Path,
+			c.nsOpts.withCapabilityPortMap(portMaps.ports),
+			c.nsOpts.withArgs(cniArgs),
+		)
+		// return early: either the netns is ok or we need to start the alloc
+		// over anyways
+		return nil, err
```

Comment:
Note to reviewers: the CNI 1.2.0 version requirement on this makes me hesitate. We've definitely run into folks who are getting older CNI plugins from their package manager and then have a bad time when we added it as a constraint. Maybe this needs a toggle to opt out of the behavior for folks who don't expect to need it? Ex. if you've got so-called immutable clients then you'll likely never need this feature.

Reply:
This is an interesting one. v1.2.0 was released almost 2 years ago, but it does seem like a lot of OS package managers have significantly older versions (not hugely surprising), so does seem like something to cover. That being said, I am wary of adding too many toggles. It's a shame we can't get the client CNI version information and make it available to the client in a useful way.

Quick check on an Ubuntu 24.04 machine:
```console
$ sudo apt-cache madison containernetworking-plugins
containernetworking-plugins | 1.1.1+ds1-3ubuntu0.24.04.2 | http://eu-west-2.ec2.archive.ubuntu.com/ubuntu noble-updates/universe amd64 Packages
containernetworking-plugins | 1.1.1+ds1-3ubuntu0.24.04.2 | http://security.ubuntu.com/ubuntu noble-security/universe amd64 Packages
containernetworking-plugins | 1.1.1+ds1-3build1 | http://eu-west-2.ec2.archive.ubuntu.com/ubuntu noble/universe amd64 Packages
```

Reply:
> It's a shame we can't get the client CNI version information and make it available to the client in a useful way.

:facepalm: We can -- we already fingerprint the CNI plugin versions! So we could gate this check behind the plugin version, and then people will get the improved behavior once they've upgraded their CNI plugins. (And we can document that we recommend using more recent versions.)